### DISCUSSION TITLE
Missing MARC data in record

### DISCUSSION THREAD

**Discussion by bjarman** [vufind-org/vufind#3999] [Q&A]
Hi!
I am evaluating VuFind and have recently made a small marc import. The import seemed to have worked. However when searching for some posts I get this error when clicking the suggestion below the search field: "Missing MARC data in record 4ncptnw32mqktrx9". The error is in file MarcReaderTrait.php line 85:
if (empty($this->fields[$preferredMarcField])) {
    throw new \Exception('Missing MARC data in record ' . $this->getUniqueId());
}
Some imported posts works but some does not. 
The imported marc posts come from the National Library in Sweden, Libris and all have a unique id in field 001. 

So, my question is, has anyone else experienced this issue?

Posts exists and are found by the auto suggest but generates error when clicking the suggestion below the search input.

**Comment by demiankatz** [comment]
@bjarman, what process did you use to create your index? This error seems to suggest that the MARC data is missing from the Solr index, so understanding your ingest process will help with troubleshooting. If you could share your import configuration and/or some sample records that are causing problems, that might also be helpful.

**Comment by bjarman** [comment]
Hi! I used the bundled import/index script in VuFind. I have made some further analysis and have found, perhaps, a clue. All imported/indexed marc records that have the field 010 are working fine. All records, that I have tried so far, that does not work do not have the field 010. My suspicion is that the field 010 must be present in the marc records?
I have not made any changes to any import configuration. 

Here is a working record:

=LDR  01528cam a22004097a 4500
=001  9325568
=003  LIBRIS
=005  20241004140612.0
=008  040114s2004\\\\xxua|||||b||||001\0|eng|c
=010  \\$a2004000097
=020  \\$a0195162226$qinb.
=020  \\$a0195162234$qhft.
=035  \\$a(DLC)  2004000097
=040  \\$aDLC$cDLC$dU$ea
=041  \\$aeng
=042  \\$apcc
=043  \\$af-ua---
=050  \0$aBR190
=082  00$a271/.009/015$222
=084  \\$aCda$2kssb/7
=084  \\$aCj.2$2kssb/7
=084  \\$aCkt:k.2$2kssb/7
=100  1\$aHarmless, William,$cS.  J.,$d1953-$4aut
=245  10$aDesert Christians :$ban introduction to the literature of early monasticism /$cWilliam Harmless.
=264  \1$aNew York :$bOxford University Press,$c2004
=300  \\$axxiv, 488 s.$bill.
=650  \7$aÖkenfäder$0https://id.kb.se/term/sao/%C3%96kenf%C3%A4der$2sao
=650  \0$aDesert Fathers
=650  \0$aChristian literature, Early$xHistory and criticism.
=650  \0$aMonastic and religious life$xHistory$yEarly church, ca. 30-600.
=887  \\$a{"@id":"xf76fm081lhjps6","modified":"2024-10-04T14:06:12.888+02:00"}$2librisxl
=976  \0$aCda$bKyrkofädernas skrifter och fornkristen litteratur
=976  \0$aCj.2$bKyrkohistoria: forntiden
=976  \0$aCkt:k.2$bAndliga ordnar och klosterväsen: historia: forntiden
=841  \\$5U$ax  a$b0408114|00008||||001|||||0901128$eu
=852  \\$5U$bU$hÖppen samling Cj$j473
=887  \\$5U$a{"@id":"48f67xb94q98xlh","modified":"2004-08-11T16:46:15+02:00"}$2librisxl

Here is a not working record:

=LDR  01895cam a22004337i 4500
=001  4ncptnw32mqktrx9
=003  LIBRIS
=005  20241002124740.0
=008  240412s2024\\\\sw\|||||||||||000\f|swe|d
=020  \\$a9789180505536$qinbunden
=035  \\$a(BOKR)9789180505536
=040  \\$aBOKR$bswe$dVasb$ei$erda
=041  \\$aswe
=082  04$a839.738$223/swe
=084  \\$aHc.01$2kssb/8
=100  1\$aWigh, Helén,$d1984-$4aut$0https://libris.kb.se/hwhsw02ffhsfk5cj#it
=245  10$aRista :$ben Elin Roth-roman /$cHelén Wigh.
=264  \1$aLund :$bHistoriska media,$c[2024]
=264  \3$aFalun :$bScandbook,$c2024
=300  \\$a382 sidor$c22 cm
=336  \\$atext$btxt$2rdacontent
=337  \\$aunmediated$bn$2rdamedia
=338  \\$avolume$bnc$2rdacarrier
=500  \\$aOriginalupplaga 2024
=600  14$aRoth, Elin$c(fiktiv gestalt)
=648  \7$a1950-talet$0https://id.kb.se/term/sao/1950-talet$2sao
=650  \7$aByar$0https://id.kb.se/term/sao/Byar$2sao
=650  \7$aEfterforskningar$0https://id.kb.se/term/sao/Efterforskningar$2sao
=650  \7$aHemligheter$0https://id.kb.se/term/sao/Hemligheter$2sao
=651  \7$aSverige$zVärmland$0https://id.kb.se/term/sao/Sverige--V%C3%A4rmland$2sao
=655  \7$aHistoriska skildringar$0https://id.kb.se/term/saogf/Historiska%20skildringar$2saogf
=655  \7$aThrillers$0https://id.kb.se/term/saogf/Thrillers$2saogf
=655  \7$aRomaner$0https://id.kb.se/term/saogf/Romaner$2saogf
=887  \\$a{"@id":"4ncptnw32mqktrx9","modified":"2024-10-02T12:47:40.227+02:00"}$2librisxl
=976  \0$aHc.01$bSvensk skönlitteratur: romaner och noveller
=841  \\$5U$axr a$b241004||0000|||||000||||||000000$eu
=852  \\$5U$bU$hMagasin H$j55121$xplikt241004
=949  \\$5U$Dcarmaglsl$F1$Gcarmaglsl$Xdisklan$616000003197518$aMagasin H 55121
=887  \\$5U$a{"@id":"4p2gk8fq2swkw2v7","modified":"2024-10-04T11:25:22.543+02:00"}$2librisxl

Do I perhaps need to make my own local marc config - omitting field 010 in some way?

**Comment by demiankatz** [comment]
@bjarman, I compiled the provided sample records into binary format and loaded them with SolrMarc in my test environment:

```
dkatz@dkatz-VirtualBox:/usr/local/vufind$ ./import-marc.sh /code/test.mrc 
Now Importing  /code/test.mrc ...
Oct 09, 10:48:32 /usr/lib/jvm/default-java/bin/java -Xms512m -Xmx512m -DentityExpansionLimit=0 -Duser.timezone=UTC -Dlog4j.configuration=file:///usr/local/vufind/import/log4j.properties  -jar /usr/local/vufind/import/solrmarc_core_3.5.jar /usr/local/vufind/local/import/import.properties -solrj /usr/local/vufind/solr/vendor/.solrj -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib  /code/test.mrc
INFO [main] (Boot.java:47) - Starting SolrMarc boot loader shim
INFO [main] (LoggerDelegator.java:272) - Command line: 
INFO [main] (LoggerDelegator.java:272) - java -Xms512m
INFO [main] (LoggerDelegator.java:272) -     -Xmx512m
INFO [main] (LoggerDelegator.java:272) -     -DentityExpansionLimit=0
INFO [main] (LoggerDelegator.java:272) -     -Duser.timezone=UTC
INFO [main] (LoggerDelegator.java:272) -     -Dlog4j.configuration=file:///usr/local/vufind/import/log4j.properties
INFO [main] (LoggerDelegator.java:272) -     -classpath /usr/local/vufind/import/solrmarc_core_3.5.jar
INFO [main] (LoggerDelegator.java:272) -     /usr/local/vufind/import/solrmarc_core_3.5.jar
INFO [main] (LoggerDelegator.java:272) -         /usr/local/vufind/local/import/import.properties
INFO [main] (LoggerDelegator.java:272) -         -solrj /usr/local/vufind/solr/vendor/.solrj
INFO [main] (LoggerDelegator.java:272) -         -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib /code/test.mrc
INFO [main] (LoggerDelegator.java:272) - Invoking main program: 
INFO [main] (LoggerDelegator.java:272) - java -classpath <CLASSPATH>
INFO [main] (LoggerDelegator.java:272) -         org.solrmarc.driver.ConfigDriver
INFO [main] (LoggerDelegator.java:272) -         /usr/local/vufind/local/import/import.properties
INFO [main] (LoggerDelegator.java:272) -         -solrj /usr/local/vufind/solr/vendor/.solrj
INFO [main] (LoggerDelegator.java:272) -         -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib
INFO [main] (LoggerDelegator.java:272) -         /code/test.mrc
DEBUG [main] (ConfigDriver.java:25) - Using config /usr/local/vufind/local/import/import.properties to initialize SolrMarc
INFO [main] (ConfigDriver.java:133) - Effective Command Line is:
INFO [main] (ConfigDriver.java:134) -    java -jar solrmarc_core.jar IndexDriver -reader_opts import.properties -dir /usr/local/vufind/local/import|/usr/local/vufind/import;local/import -config "marc.properties, marc_local.properties" -solrURL http://localhost:8983/solr/biblio/update -solrj /usr/local/vufind/solr/vendor/.solrj -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib /code/test.mrc
INFO [main] (LoggerDelegator.java:272) - Invoking main program: 
INFO [main] (LoggerDelegator.java:272) - java -classpath <CLASSPATH>
INFO [main] (LoggerDelegator.java:272) -         org.solrmarc.driver.IndexDriver
INFO [main] (LoggerDelegator.java:272) -         -reader_opts import.properties
INFO [main] (LoggerDelegator.java:272) -         -dir /usr/local/vufind/local/import|/usr/local/vufind/import;local/import
INFO [main] (LoggerDelegator.java:272) -         -config marc.properties, marc_local.properties
INFO [main] (LoggerDelegator.java:272) -         -solrURL http://localhost:8983/solr/biblio/update
INFO [main] (LoggerDelegator.java:272) -         -solrj /usr/local/vufind/solr/vendor/.solrj
INFO [main] (LoggerDelegator.java:272) -         -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib
INFO [main] (LoggerDelegator.java:272) -         /code/test.mrc
DEBUG [main] (BootableMain.java:119) - Adding directory: /usr/local/vufind/local/import
DEBUG [main] (BootableMain.java:119) - Adding directory: /usr/local/vufind/import
DEBUG [main] (Boot.java:713) - Adding jars files in directory: /usr/local/vufind/import/lib_local
DEBUG [main] (Boot.java:713) - Adding jars files in directory: /usr/local/vufind/solr/vendor/modules/analysis-extras/lib
 INFO [main] (ValueIndexerFactory.java:119) - Using directory: /usr/local/vufind/import/index_java as location of java sources
 INFO [main] (PropertyUtils.java:337) - Opening file (instead of 1 other options): /usr/local/vufind/local/import/import.properties
DEBUG [main] (SolrCoreLoader.java:48) - Pinging Solr at URL:  http://localhost:8983/solr/biblio/admin/ping
DEBUG [main] (SolrCoreLoader.java:65) -     ...
DEBUG [main] (SolrCoreLoader.java:54) -       "status":"OK"
DEBUG [main] (SolrCoreLoader.java:96) - Found Solrj class org.apache.solr.client.solrj.impl.HttpSolrClient$Builder
 INFO [main] (IndexDriver.java:166) - Reading and compiling index specifications: marc.properties, marc_local.properties
 INFO [main] (IndexDriver.java:257) - Opening index spec file: /usr/local/vufind/import/marc.properties
 INFO [main] (IndexDriver.java:257) - Opening index spec file: /usr/local/vufind/import/marc_local.properties
 INFO [main] (IndexDriver.java:100) - Opening input files: [/code/test.mrc]
DEBUG [main] (IndexDriver.java:331) - System Class Path = /usr/local/vufind/import/solrmarc_core_3.5.jar
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 325568
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : ncptnw32mqktrx9
 INFO [main] (ThreadedIndexer.java:259) - Done with all indexing, finishing writing records to solr
 INFO [main] (ThreadedIndexer.java:272) - Done writing records to solr
 INFO [main] (Indexer.java:595) - Commmiting updates to Solr
 INFO [main] (IndexDriver.java:391) - 2 records read
 INFO [main] (IndexDriver.java:392) - 2 records indexed  and 
 INFO [main] (IndexDriver.java:399) - 2 records sent to Solr in 1.12 seconds
```

Both records display correctly:

![image](https://github.com/user-attachments/assets/a60313a1-a04d-497f-a663-33ae93f6ef9f)

![image](https://github.com/user-attachments/assets/8a783642-49a5-404d-8703-f68f93d7c93b)

I don't think the issue is likely to be related to the 010 field. When SolrMarc indexes a record, it not only maps selected data to specialized Solr fields, it also is supposed to store a representation of the full record in a field called `fullrecord`. Your problem seems to be caused by this field being unpopulated. That should never happen under normal circumstances, so either there's a misconfiguration, or there's something about the records that is preventing them from encoding correctly. I'd be interested to know whether there are any errors in the output of your import command, for example. As you can see above, everything worked fine for me, but since I've rebuilt the records, it's possible if there's an encoding problem that it got worked out through all the conversion steps. :-)

**Comment by bjarman** [comment]
I will check the logs from my import tomorrow and get back. Thank you so
much so far!

ons 9 okt. 2024 kl. 16:52 skrev Demian Katz ***@***.***>:

**Comment by bjarman** [comment]
Ok, I have found some warnings in the import log. There are a lot of warnings relating to FullRecordAsJson.
Most of them looks like this:
2024-10-09 12:26:19,089 [RecordIndexer-Thread-0] WARN  RecordIndexer-Thread-0 org.solrmarc.driver.Indexer - while processing index specification:  FullRecordAsJSON2
2024-10-09 12:26:19,091 [RecordIndexer-Thread-0] INFO  RecordIndexer-Thread-0 org.vufind.index.CreatorTools - Unknown relator: 9pu
2024-10-09 12:26:19,224 [RecordIndexer-Thread-0] WARN  RecordIndexer-Thread-0 org.solrmarc.driver.Indexer - while processing index specification:  FullRecordAsJSON2
2024-10-09 12:26:19,231 [RecordIndexer-Thread-0] INFO  RecordIndexer-Thread-0 org.vufind.index.CreatorTools - Unknown relator: ed.

Does this shine a light on what I might be doing wrong?
Is there a way to delete al imported records and re-import them correctly?

**Reply by demiankatz** [reply]
Okay, that's very helpful -- we have a step that formats the input MARC into MARC-in-JSON format for storage in the index, and for some reason, some of your records are causing that conversion to fail. It would be very helpful if you could share a problem record in the original format so we can analyze what is going wrong, since it seems that whatever the problem is, it gets worked out if we convert the record into different intermediate formats.

In the meantime, a simple workaround is to store the MARC in a different format, since VuFind supports MARC-in-JSON, MARCXML and binary MARC (we choose the JSON option by default because it's a good compromise between the verbosity of XML and the baked-in length limits of the binary format). To choose a different option, just copy $VUFIND_HOME/import/marc_local.properties into $VUFIND_LOCAL_DIR/import/ (if you haven't already) and then uncomment one of the alternate formatting options here:

https://github.com/vufind-org/vufind/blob/dev/import/marc_local.properties#L12-L17

You can simply reindex the same file, and the bad records should get overwritten with the corrected versions.

If you ever need to fully erase and rebuild your index, there is a [Re-Indexing wiki page](https://vufind.org/wiki/indexing:re-indexing) with details... but when you index a record, Solr just replaces any existing record with the same ID, so as long as you're re-indexing exactly the same set of records (as you are in this case), there's no need to take extra steps. Wiping the index is generally only necessary if the wrong IDs were indexed in the first place.

I hope this helps!

**Comment by bjarman** [comment]
The file I used is attached here.
[UUB20241005.mrc.zip](https://github.com/user-attachments/files/17327486/UUB20241005.mrc.zip)

**Reply by demiankatz** [reply]
Yes, I am seeing the same problem that you encountered when I use this in my test environment. What process did you use to create/obtain this MARC file? What encoding is used? I tried to use yaz-marcdump to convert it to XML as an experiment, but that also failed, with this output:

```
<!-- Skipping bad byte 61 (0x3D) at offset 0 (0x0) -->
<!-- Skipping bad byte 76 (0x4C) at offset 1 (0x1) -->
<!-- Skipping bad byte 68 (0x44) at offset 2 (0x2) -->
<!-- Skipping bad byte 82 (0x52) at offset 3 (0x3) -->
<!-- Skipping bad byte 32 (0x20) at offset 4 (0x4) -->
<!-- Skipping bad byte 32 (0x20) at offset 5 (0x5) -->
<!-- EOF while searching for RS -->
```

I suspect the problem is that either the input file is being generated incorrectly, or it is in an unusual encoding that will require special configuration for correct processing.

**Reply by demiankatz** [reply]
Ohh, wait, I see! The SolrMarc tool expects either binary MARC or MARC-XML as an input. This is in pure text format, which is not supported. I'm surprised it works at all; apparently there is some limited support for this file format, but since it is not officially supported, it's only partially working. If you can compile this to binary or XML format, then I expect that everything should work correctly for you. I don't think yaz-marcdump supports this format either, but you can use MarcEdit to do the conversion.

Please let me know if I can be of any further assistance!

**Reply by bjarman** [reply]
It is exported from the national library in Sweden. From what I can see when looking at the file in Visual Studio Code it is UTF-8?
![image](https://github.com/user-attachments/assets/1ace094a-051f-4205-bb47-0f693e098b9c)

**Reply by bjarman** [reply]
Aha! I will search for MarcEdit and try to compile it!

**Reply by demiankatz** [reply]
What export tool did you use? Does it support other MARC carrier formats? The "=LDR ..." format you are using here is commonly used for human-readable transmission, but VuFind's tools expect one of the machine-readable formats. As mentioned above, the only tool I'm personally aware of that supports this format is MarcEdit, which you can use to compile the text into binary MARC that VuFind can understand.

**Reply by demiankatz** [reply]
You can find MarcEdit here: https://marcedit.reeset.net/downloads

**Reply by demiankatz** [reply]
Also, for the record, I noted this situation in an existing open issue on the SolrMarc project describing possible problems with the JSON2 conversion: https://github.com/solrmarc/solrmarc/issues/103

**Comment by bjarman** [comment]
Ok, we exported to marc xml and I redid the import. The problem seems to persist. This is the log from the import: 
[root@uub-sll-vufind vufind]# ./import-marc.sh import/testimport.xml
Now Importing  /usr/local/vufind/import/testimport.xml ...
Oct 10, 14:46:31 java -Xms512m -Xmx512m -DentityExpansionLimit=0 -Duser.timezone=UTC -Dlog4j.configuration=file:///usr/local/vufind/import/log4j.properties  -jar /usr/local/vufind/import/solrmarc_core_3.5.jar /usr/local/vufind/local/import/import.properties -solrj /usr/local/vufind/solr/vendor/.solrj -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib  /usr/local/vufind/import/testimport.xml
INFO [main] (Boot.java:47) - Starting SolrMarc boot loader shim
INFO [main] (LoggerDelegator.java:272) - Command line:
INFO [main] (LoggerDelegator.java:272) - java -Xms512m
INFO [main] (LoggerDelegator.java:272) -     -Xmx512m
INFO [main] (LoggerDelegator.java:272) -     -DentityExpansionLimit=0
INFO [main] (LoggerDelegator.java:272) -     -Duser.timezone=UTC
INFO [main] (LoggerDelegator.java:272) -     -Dlog4j.configuration=file:///usr/local/vufind/import/log4j.properties
INFO [main] (LoggerDelegator.java:272) -     -classpath /usr/local/vufind/import/solrmarc_core_3.5.jar
INFO [main] (LoggerDelegator.java:272) -     /usr/local/vufind/import/solrmarc_core_3.5.jar
INFO [main] (LoggerDelegator.java:272) -         /usr/local/vufind/local/import/import.properties
INFO [main] (LoggerDelegator.java:272) -         -solrj /usr/local/vufind/solr/vendor/.solrj
INFO [main] (LoggerDelegator.java:272) -         -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib /usr/local/vufind/import/testimport.xml
INFO [main] (LoggerDelegator.java:272) - Invoking main program:
INFO [main] (LoggerDelegator.java:272) - java -classpath <CLASSPATH>
INFO [main] (LoggerDelegator.java:272) -         org.solrmarc.driver.ConfigDriver
INFO [main] (LoggerDelegator.java:272) -         /usr/local/vufind/local/import/import.properties
INFO [main] (LoggerDelegator.java:272) -         -solrj /usr/local/vufind/solr/vendor/.solrj
INFO [main] (LoggerDelegator.java:272) -         -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib
INFO [main] (LoggerDelegator.java:272) -         /usr/local/vufind/import/testimport.xml
DEBUG [main] (ConfigDriver.java:25) - Using config /usr/local/vufind/local/import/import.properties to initialize SolrMarc
INFO [main] (ConfigDriver.java:133) - Effective Command Line is:
INFO [main] (ConfigDriver.java:134) -    java -jar solrmarc_core.jar IndexDriver -reader_opts import.properties -dir /usr/local/vufind/local/import|/usr/local/vufind/import;local/import -config "marc.properties, marc_local.properties" -solrURL http://localhost:8983/solr/biblio/update -solrj /usr/local/vufind/solr/vendor/.solrj -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib /usr/local/vufind/import/testimport.xml
INFO [main] (LoggerDelegator.java:272) - Invoking main program:
INFO [main] (LoggerDelegator.java:272) - java -classpath <CLASSPATH>
INFO [main] (LoggerDelegator.java:272) -         org.solrmarc.driver.IndexDriver
INFO [main] (LoggerDelegator.java:272) -         -reader_opts import.properties
INFO [main] (LoggerDelegator.java:272) -         -dir /usr/local/vufind/local/import|/usr/local/vufind/import;local/import
INFO [main] (LoggerDelegator.java:272) -         -config marc.properties, marc_local.properties
INFO [main] (LoggerDelegator.java:272) -         -solrURL http://localhost:8983/solr/biblio/update
INFO [main] (LoggerDelegator.java:272) -         -solrj /usr/local/vufind/solr/vendor/.solrj
INFO [main] (LoggerDelegator.java:272) -         -lib_local /usr/local/vufind/import/lib_local;/usr/local/vufind/solr/vendor/modules/analysis-extras/lib
INFO [main] (LoggerDelegator.java:272) -         /usr/local/vufind/import/testimport.xml
DEBUG [main] (BootableMain.java:119) - Adding directory: /usr/local/vufind/local/import
DEBUG [main] (BootableMain.java:119) - Adding directory: /usr/local/vufind/import
DEBUG [main] (Boot.java:713) - Adding jars files in directory: /usr/local/vufind/import/lib_local
DEBUG [main] (Boot.java:713) - Adding jars files in directory: /usr/local/vufind/solr/vendor/modules/analysis-extras/lib
 INFO [main] (ValueIndexerFactory.java:119) - Using directory: /usr/local/vufind/import/index_java as location of java sources
 INFO [main] (PropertyUtils.java:337) - Opening file (instead of 1 other options): /usr/local/vufind/local/import/import.properties
DEBUG [main] (SolrCoreLoader.java:48) - Pinging Solr at URL:  http://localhost:8983/solr/biblio/admin/ping
DEBUG [main] (SolrCoreLoader.java:65) -     ...
DEBUG [main] (SolrCoreLoader.java:54) -       "status":"OK"
DEBUG [main] (SolrCoreLoader.java:96) - Found Solrj class org.apache.solr.client.solrj.impl.HttpSolrClient$Builder
 INFO [main] (IndexDriver.java:166) - Reading and compiling index specifications: marc.properties, marc_local.properties
 INFO [main] (IndexDriver.java:257) - Opening index spec file: /usr/local/vufind/import/marc.properties
 INFO [main] (IndexDriver.java:257) - Opening index spec file: /usr/local/vufind/import/marc_local.properties
 INFO [main] (IndexDriver.java:100) - Opening input files: [/usr/local/vufind/import/testimport.xml]
DEBUG [main] (IndexDriver.java:331) - System Class Path = /usr/local/vufind/import/solrmarc_core_3.5.jar
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 78111
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 164784
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 1161871
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 1174369
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 1275546
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 1282100
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 1293723
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 1369353
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 1392429
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 2108262
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 2201823
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 2231220
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 7228637
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 7265767
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 7270484
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 7274515
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 7279044
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 7658614
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 7771704
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 8217300
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 8218246
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 8226389
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 20057490
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 8381103
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 8408105
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 8954290
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 12746573
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 9174541
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 10058625
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 10164938
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 10201208
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 9665574
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 9665578
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 11810929
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 13534721
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 19375169
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 21656347
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : n0jx8m2wlckm72t3
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 17793233
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 18096202
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 19659284
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 18xfp0hhzr73tvds
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 08h658mmxg7gkncv
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 3ftcvnxr1dv7hrch
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 3kk7d8d11frhjb1w
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : jx7pg50kgznpwt8q
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : k11twcq6hs6k66fq
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : hzbwfj2kf5bt56g1
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : q6k4nrbxn8g96jbk
DEBUG [MarcReader-Thread] (MarcReaderThread.java:35) - record read : 6p055dz34ddqx7nl
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: 1174369
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: 1282100
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: 1369353
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: 20057490
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: 12746573
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: 10164938
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: 21656347
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: 19659284
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 WARN [RecordIndexer-Thread-0] (Indexer.java:455) - Exception in record: q6k4nrbxn8g96jbk
 WARN [RecordIndexer-Thread-0] (Indexer.java:456) - while processing index specification:  FullRecordAsJSON2
 INFO [main] (ThreadedIndexer.java:259) - Done with all indexing, finishing writing records to solr
 INFO [main] (ThreadedIndexer.java:272) - Done writing records to solr
 INFO [main] (Indexer.java:595) - Commmiting updates to Solr
 INFO [main] (IndexDriver.java:391) - 50 records read
 INFO [main] (IndexDriver.java:392) - 50 records indexed  and
 INFO [main] (IndexDriver.java:399) - 50 records sent to Solr in 4.74 seconds

**Comment by bjarman** [comment]
And here is the xml-file I used
[testimport.xml.zip](https://github.com/user-attachments/files/17328447/testimport.xml.zip)

**Reply by demiankatz** [reply]
Okay, very interesting! It looks like I'm falsely blaming the carrier format for the problem. It seems to go back to my [earlier comment](https://github.com/vufind-org/vufind/discussions/3999#discussioncomment-10902832) -- the problem is that there's something in the records that the very strict JSON2 conversion doesn't like, but if you switch to a different storage format for the Solr index, the process works correctly. For example, in my test environment, I switched to use XML storage by uncommenting `fullrecord = FullRecordAsXML` in my marc_local.properties file, and then the ingest succeeded (even using your original text file).

The issue is likely related to the open SolrMarc issue (https://github.com/solrmarc/solrmarc/issues/103) that I cited earlier. In the meantime, simply using a different storage format in your index should be a suitable workaround. The only consequence of MARC XML storage is that your index size will be slightly larger because it's a less efficient encoding... but that's unlikely to cause any issues for you.

**Comment by bjarman** [comment]
Ok! Thank you so much for investigating this! I will switch storage format
to XML and hopefully it will work then!

I am unable to do this right now due to travel. I will let you know as soon
as it is done though.

Have a great weekend!

tors 10 okt. 2024 kl. 15:34 skrev Demian Katz ***@***.***>:

**Reply by demiankatz** [reply]
You too, and safe travels. :-)

**Comment by bjarman** [comment]
Hi! 
I uncommented fullrecord = FullRecordAsXML in my in my marc_local.properties file and reimported the marc-xml file. This import went fine. However I got the same errors as before. So I redid the import using the mrc file and this time everything seems to be working - although I still got some errors in the import log (Unknown relator: 9pu and Unknown relator: ed.). No errors when searching for records or using links in the interface. 

I have no idea why this worked but thank you either way :) 

Unless you want me to test/try to get the import using the marc-xml file working I guess we can close this.

**Reply by demiankatz** [reply]
Hmm, it's strange that the XML version didn't work, but maybe MarcEdit did something wrong in terms of encoding, etc. If the original file works for you, that's easier anyway, so I don't think it's worth spending more time troubleshooting what happened with the XML version.

It's generally safe to ignore those "unknown relator" messages -- this is just informing you that the relator code is not found in author-classification.ini. It will still be displayed and indexed, but if you were trying to standardize everything for translation purposes, this code is not on the list of recognized values and will likely not be internationalized by default.

I think it's safe to close this if you're satisfied that we've solved all of your outstanding problems!

**Comment by he247** [comment]
Dear @demiankatz  and @bjarman  did you already find a solution to this? I have several records that seem to be problematic, when using FullRecordAsJSON2.

**Reply by demiankatz** [reply]
I think a full solution would require upstream changes to Marc4j and a new release of SolrMarc; as far as I know, that has not happened yet.

In the meantime, I am still using FullRecordAsJSON2, but I have added a test in my Nagios monitoring system to query Solr for any records with values in the marc_error field and trigger an alert when the result is non-empty. This doesn't solve the problem, but it immediately alerts me when new records are imported with problems, and then I can correct them. This has the advantage of alerting us to bad records as they come in, rather than silently ignoring them and allowing them to potentially confuse our users. (Usually when something goes wrong, it has to do with character encoding issues, and the resulting record is completely unreadable without human intervention).
